File: src/model/QrAccount.ts
import { PublicKey } from '@solana/web3.js';
import BN from 'bn.js';

export type QrAccountWithPubkey = {
  publicKey: PublicKey;
  data: QrAccountData;
};

export type QrAccountData = {
  amount: BN;
  authority: PublicKey;
  bump: number;
  hash: string;
  lastTransferTimestamp: BN;
  tokens: PublicKey[];
  tokensStats: {
    transferCount: BN;
    totalAmount: BN;
    totalValue: BN;
  }[];
};

export type QrAccountJsonWithPubkey = {
  publicKey: string;
  data: QrAccountJson;
};

export type QrAccountJson = {
  amount: string;
  authority: string;
  bump: number;
  hash: string;
  lastTransferTimestamp: string;
  tokens: string[];
  tokensStats: {
    transferCount: string;
    totalAmount: string;
    totalValue: string;
  }[];
};

export class QrAccount {
  static fromJsonWithPubkey(
    json: QrAccountJsonWithPubkey
  ): QrAccountWithPubkey {
    return {
      publicKey: new PublicKey(json.publicKey),
      data: QrAccount.fromJson(json.data),
    };
  }

  static fromJson(json: QrAccountJson): QrAccountData {
    return {
      amount: new BN(json.amount),
      authority: new PublicKey(json.authority),
      bump: json.bump,
      hash: json.hash,
      lastTransferTimestamp: new BN(json.lastTransferTimestamp),
      tokens: json.tokens.map((token) => new PublicKey(token)),
      tokensStats: json.tokensStats.map((stats) => ({
        transferCount: new BN(stats.transferCount),
        totalAmount: new BN(stats.totalAmount),
        totalValue: new BN(stats.totalValue),
      })),
    };
  }

  static toJsonWithPubkey(
    qrAccount: QrAccountWithPubkey
  ): QrAccountJsonWithPubkey {
    return {
      publicKey: qrAccount.publicKey.toString(),
      data: QrAccount.toJson(qrAccount.data),
    };
  }

  static toJson(qrAccount: QrAccountData): QrAccountJson {
    return {
      amount: qrAccount.amount.toString(),
      authority: qrAccount.authority.toBase58(),
      bump: qrAccount.bump,
      hash: qrAccount.hash,
      lastTransferTimestamp: qrAccount.lastTransferTimestamp.toString(),
      tokens: qrAccount.tokens.map((token) => token.toBase58()),
      tokensStats: qrAccount.tokensStats.map((stats) => ({
        transferCount: stats.transferCount.toString(),
        totalAmount: stats.totalAmount.toString(),
        totalValue: stats.totalValue.toString(),
      })),
    };
  }
}

File: contract/programs/hemmorphi/src/state/qr_account.rs
use anchor_lang::prelude::*;
use anchor_lang::{prelude::Pubkey, InitSpace};

#[account]
#[derive(InitSpace)]
pub struct QrAccount {
    pub authority: Pubkey,            // Owner of the QR code
    pub amount: u64,                  // Value associated with the QR code
    pub last_transfer_timestamp: i64, // Timestamp of the last scan
    pub bump: u8,                     // PDA bump seed
    #[max_len(5, 32)]
    pub tokens: Vec<Pubkey>, // List of tokens
    #[max_len(5)]
    pub tokens_stats: Vec<TokenStats>, // Stats for each token
    #[max_len(32)]
    pub hash: String, // Unique identifier
}

impl QrAccount {
    pub const TOKENS_MAX_COUNT: usize = 5;

    pub fn check_if_token_exists(&self, token: Pubkey) -> bool {
        self.tokens.iter().any(|t| *t == token)
    }

    pub fn check_if_correct_amount(&self, amount: u64) -> bool {
        if self.amount == 0 {
            return true;
        } else {
            return self.amount == amount;
        }
    }

    pub fn update_token_stats(&mut self, token: Pubkey, amount: u64) {
        let index = self.tokens.iter().position(|t| *t == token).unwrap();
        self.tokens_stats[index].transfer_count += 1;
        self.tokens_stats[index].total_amount += amount;
        self.tokens_stats[index].total_value += amount;
    }
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Default, InitSpace)]
pub struct TokenStats {
    pub transfer_count: u64, // Number of transfers
    pub total_amount: u64,   // Total amount transferred
    pub total_value: u64,    // Total value transferred
}
